proposals/time-stamp-counter-support.md

```diff
@@ -0,0 +1,274 @@
+implementations for atomics, including atomics which may not exactly fit under
+the C11 memory model. This module SHOULD NOT be used for anything which can be
+implemented on other architectures, for instance, x86's `popcnt` can be
+implemented as `vcnt.8` with ARM NEON or with a for loop, so it shouldn't be
+there. In other words, every single function in this module and it's submodules
+will have a constraint requiring a specific architecture.
+
+As we want Mojo to be as portable as possible, the module shall be hidden by
+default and the documentation of this module should discourage end-users using
+its functions unless there are no other options.
+
+### The creation of time._tsc
+
+The time module is the most logical place for most of this functionality to
+live, since they are time stamp counters. In this module, different
+architectures will be unified into a single portable API consisting of the
+following:
+
+```mojo
+@value
+struct TscFrequency:
+    """
+    Contains the TSC frequency with architecture-defined units. 
+    """
+    var _tsc_frequency: UInt64
+
+    fn get_frequency_hz(borrowed self) -> UInt64:
+        """
+        Get the frequency in Hertz. 
+        """
+        pass
+
+@value
+struct TscTimestamp:
+    """
+    Encapsulates the value of a TSC read. Units are architecture dependent.
+
+    Implementation detail: Can only reliably hold up to 2^64-1 nanoseconds or ~580 years on most architectures. 
+    """
+    var _tsc_value: UInt64
+
+    fn time_since_ns(borrowed self, borrowed other: Self, borrowed freq: TscFrequency) -> Int64:
+        """
+        Returns the time since the previous timestamp in nanoseconds using the provided frequency reference.
+        """
+        ...
+
+    fn get_raw(borrowed self) -> UInt64:
+        """
+        Access the raw, architecture dependent value of the TSC. 
+        """
+        return self._tsc_value
+
+struct Tsc:
+    """
+    Time Stamp Counter (TSC) container to hold the expensive initialization of 
+    the TSC API, and also allows for converting tsc timestamps into values 
+    usable by other APIs. Treats start_os_timestamp and start_tsc_timestamp as
+    referring to the same instant in time.
+
+    WARNING: There are hazards to the use of this API, please read the module
+    documentation, specifically the "Limitations, Hazards and Cautions" section.
+
+    freq: The frequency of the TSC, and where most of the expensive 
+    initialization comes from.
+
+    start_os_timestamp: Number of nanoseconds since an undefined point in time 
+    (typically the platform epoch). Used to aid in converting TscTimestamp 
+    instances to timestamps
+
+    start_tsc_timestamp: A tsc timestamp taken right after the timestamp 
+    syscall call returned. 
+
+    tsc_is_reliable: Whether the TSC is usable as a time source. If false, a monotonic but fast to query OS clock is used.
+    """
+
+    var freq: TscFrequency
```

Comment:
**Suggestion** Have you considered lifting the frequency and `tsc_is_reliable` out of this struct?  This will be the de-facto type passed around, so we'd want to be mindful of size and layout with respect to properties that conceptually are static for a system and don't differ between two different timestamp instances.

Reply:
The frequency is the most expensive value to determine on some architectures:

| Arch | Cost |
|--------|--------|
| aarch64 | MSR Read |
| RISC-V | MSR Read |
| AMD x86-64 | At least 1 syscall, parsing an int from a string (Linux) |
| Intel x86-64 | Traversal through CPUID and a variety of lookup tables |

There are two ways I can lift it out to a global. The first is to just store one, which will break ancient x86 CPUs (pre-
Nehalem on Intel, and I'm fairly sure pre-Zen for AMD), due to drift between cores/sockets, on ARM the architecture requires a single clock, so that's fine, and RISC-V is similar for most implementations. The other option is to store a frequency for each CPU core on x86, which also means every Mojo program has to move itself between CPU cores during runtime init. That is not very desirable, so if dropping very old CPUs is fine I'll go with using a global variable for frequency, a base TSC timestamp and a WALL time. 

If I do that and have the runtime do the reliability check before main (to allow for forcing it off using an environment variable if someone's environment doesn't have good TSC), then these can all become free-standing functions and the `Tsc` struct goes away.

Reply:
It's fine to drop very old CPUs, so I'd go recommend we go that direction!

Can you help me understand more about needing to do the reliability check all **before main**?  Why does it need to happen before main and not, say, at the very beginning?

Reply:
> Can you help me understand more about needing to do the reliability check all **before main**? Why does it need to happen before main and not, say, at the very beginning?

The idea was to set the "is_tsc_reliable" flag before the user has a chance to run code that would read it to avoid races. The main downside of this approach is that under some circumstances I have to fall back to "calibrate using the system timer" to determine what the frequency of the TSC is. That involves a sleep (probably up to 100ms) and some syscalls. 

I could also it so that there's a global tsc_frequency (assuming that `KGEN_CompilerRT_GetGlobalOrNull` is fairly cheap) that is left as 0 if the TSC is uninitalized and 1 if the TSC is unreliable (since a 1 Hz timer is useless anyway) and initialize it on first use along with providing a memory fence, then this should be fine if it's only used when converting a TSC timestamp to real time. 

```mojo
fn get_tsc_freq_hz() -> UInt64:
    var tsc_freq = _get_global_or_null["__tsc_frequency"]().bitcast[UInt64]()
    if tsc_freq != UnsafePointer[UInt64]():
        if tsc_freq[] > 1:
           return tsc_freq[]
        else:
          mfence()
          if tsc_freq[] > 1:
            return tsc_freq[]
          else:
              return init_tsc()
  else:
    return init_tsc()
```

If you want accelerators supported by this, then I'll stick the current target arch name on the end of `__tsc_frequency`.